### PR TITLE
GH-18487: [R] Read Text (CSV/JSON) from character vector

### LINKED ISSUE #18487: [R] Read CSV from character vector
`readr::read_csv()` lets you read in data from a character vector, useful for (e.g.) taking the results of a system call and reading it in as a data.frame. 

```Java

> readr::read_csv(c("a,b", "1,2", "3,4"))
# A tibble: 2 x 2
      a     b
  <dbl> <dbl>
1     1     2
2     3     4
```

One solution would be similar to ARROW-9235, perhaps, treating it as a textConnection. 

Another solution is to write to a tempfile.

**Reporter**: [Neal Richardson](https://issues.apache.org/jira/browse/ARROW-11441) / @nealrichardson

<sub>**Note**: *This issue was originally created as [ARROW-11441](https://issues.apache.org/jira/browse/ARROW-11441). Please see the [migration documentation](https://github.com/apache/arrow/issues/14542) for further details.*</sub>

### PR DESCRIPTION
**Pull request by eitsupi**
### Rationale for this change

Allows literal strings to be read directly through the `I()` function in the same way as the `readr::read_csv()` function.
This is useful for checking behavior without the need to create temporary files.

```r
> read_csv_arrow(I("x,y\n1,2\n3,4"))
# A tibble: 2 × 2
      x     y
  <int> <int>
1     1     2
2     3     4
```

```r
> read_csv_arrow(I(c(
        "x,y
        1,2
        3,4"
    )))
# A tibble: 2 × 2
      x     y
  <int> <int>
1     1     2
2     3     4
```

```r
> read_csv_arrow(I(c("x,y", "1,2", "3,4")))
# A tibble: 2 × 2
      x     y
  <int> <int>
1     1     2
2     3     4
```

### What changes are included in this PR?

In `read_csv_arrow` and `read_json_arrow`, if the first argument `file` inherits `AsIs` class, `file` is now interpreted as literal data.
This is consistent with the behavior of `readr::read_csv()`, which is widely used to read text files as data frames.

This is a breaking change; the behavior of wrapping a path as a string with `I()` is changed.

For example

#### readr::read_csv

```r
> readr::read_csv(I(readr::readr_example("mtcars.csv")))
Rows: 0 Columns: 1
── Column specification ────────────────────────────────────────────────────────────────────────────────────────────────
Delimiter: ","
chr (1): /usr/local/lib/R/site-library/readr/extdata/mtcars.csv

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
# A tibble: 0 × 1
# … with 1 variable:
#   /usr/local/lib/R/site-library/readr/extdata/mtcars.csv <chr>
# ℹ Use `colnames()` to see all variable names
```
#### arrow 10.01's arrow::read_csv_arrow

```r
> arrow::read_csv_arrow(I(readr::readr_example("mtcars.csv")))
# A tibble: 32 × 11
     mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear  carb
   <dbl> <int> <dbl> <int> <dbl> <dbl> <dbl> <int> <int> <int> <int>
 1  21       6  160    110  3.9   2.62  16.5     0     1     4     4
 2  21       6  160    110  3.9   2.88  17.0     0     1     4     4
 3  22.8     4  108     93  3.85  2.32  18.6     1     1     4     1
 4  21.4     6  258    110  3.08  3.22  19.4     1     0     3     1
 5  18.7     8  360    175  3.15  3.44  17.0     0     0     3     2
 6  18.1     6  225    105  2.76  3.46  20.2     1     0     3     1
 7  14.3     8  360    245  3.21  3.57  15.8     0     0     3     4
 8  24.4     4  147.    62  3.69  3.19  20       1     0     4     2
 9  22.8     4  141.    95  3.92  3.15  22.9     1     0     4     2
10  19.2     6  168.   123  3.92  3.44  18.3     1     0     4     4
# … with 22 more rows
# ℹ Use `print(n = ...)` to see more rows
```

#### This PR's arrow::read_csv_arrow

```r
> arrow::read_csv_arrow(I(readr::readr_example("mtcars.csv")))
Error:
! Invalid: CSV parse error: Empty CSV file or block: cannot infer number of columns
Run `rlang::last_error()` to see where the error occurred.
```
* Closes: #18487

### CODE REVIEW

File: r/R/json.R [hunk 1]
```diff
@@ -43,17 +43,30 @@
 #' @examplesIf arrow_with_json()
 #' tf <- tempfile()
 #' on.exit(unlink(tf))
-#' writeLines('
-#'     { "hello": 3.5, "world": false, "yo": "thing" }
-#'     { "hello": 3.25, "world": null }
-#'     { "hello": 0.0, "world": true, "yo": null }
-#'   ', tf, useBytes = TRUE)
+#' json <- '
```
Comment (eitsupi) [inline r/R/json.R:46]: lintr seems to complain about this line... (r-lib/lintr#1908)

File: r/tests/testthat/test-json.R [hunk 2]
```diff
@@ -253,3 +253,12 @@ test_that("Can read json file with list<struct<T...>> nested columns (ARROW-7740
   expected <- tibble::tibble(a = c(list(one), list(one)))
   expect_equal(read_json_arrow(tf), expected, ignore_attr = TRUE)
 })
+
+test_that("Read literal data directory", {
```
Comment (thisisnic) [inline r/tests/testthat/test-json.R:257]:
```suggestion
test_that("Read literal data directly", {
```

File: r/R/csv.R [hunk 3]
```diff
@@ -198,6 +205,14 @@ read_delim_arrow <- function(file,
     )
   }
 
+  if (inherits(file, "AsIs")) {
+    if (is.raw(file)) {
+      file <- unclass(file)
```
Comment (thisisnic) [inline r/R/csv.R:210]: Why do we need to call  `unclass()` here?

**Comment by thisisnic** [comment]
Thanks for making these changes.  A couple of suggestions, and a question which may warrant a comment addition, but otherwise this looks good to me.

File: r/R/csv.R (hunk 3 above)
Comment (eitsupi) [reply r/R/csv.R:210]: Without it, wrapping `I()` around a raw vector will cause it to fail to read.
This behavior can be seen in the current version of arrow.

```r
> "a\n1" |> charToRaw() |> arrow::read_csv_arrow()
  a
1 1

> "a\n1" |> charToRaw() |> I() |> arrow::read_csv_arrow()
Error: file must be a "InputStream"
```

On the other hand, `readr::read_csv()` can read raw vectors with or without `I()`, so I thought it necessary to unclass `AsIs` here for consistency of behavior.

```r
> "a\n1" |> charToRaw() |> readr::read_csv()
Rows: 1 Columns: 1
── Column specification ────────────────────────────────────────────────────────────────────────────────────────────────
Delimiter: ","
dbl (1): a

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
# A tibble: 1 × 1
      a
  <dbl>
1     1

> "a\n1" |> charToRaw() |> I() |> readr::read_csv()
Rows: 1 Columns: 1
── Column specification ────────────────────────────────────────────────────────────────────────────────────────────────
Delimiter: ","
dbl (1): a

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
# A tibble: 1 × 1
      a
  <dbl>
1     1
```

Perhaps we should mention this in the comments?

Comment (thisisnic) [reply r/R/csv.R:210]: Thanks for that explanation; yes, please just add a brief comment about that, as it's not immediately apparent.

**Review by thisisnic — APPROVED** [review]

**Pull request merged**